### PR TITLE
common: distance_sensor: add signal_quality field (mavlink#1403)

### PR DESCRIPTION
**Pull request by amilcarlucas**
Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

@peterbarker complained that this needs a ArduPilot code change. But the code change has already been done in https://github.com/ArduPilot/ardupilot/pull/15404/commits/55676914512577ad33e93356bcc95a6b10c975a6 and the CI tests pass

### CODE REVIEW

**Review by peterbarker — COMMENTED** [review]
Seems fine.

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -4714,6 +4714,7 @@
       <field type="float" name="horizontal_fov" units="rad">Horizontal Field of View (angle) where the distance measurement is valid and the field of view is known. Otherwise this is set to 0.</field>
       <field type="float" name="vertical_fov" units="rad">Vertical Field of View (angle) where the distance measurement is valid and the field of view is known. Otherwise this is set to 0.</field>
       <field type="float[4]" name="quaternion">Quaternion of the sensor orientation in vehicle body frame (w, x, y, z order, zero-rotation is 1, 0, 0, 0). Zero-rotation is along the vehicle body x-axis. This field is required if the orientation is set to MAV_SENSOR_ROTATION_CUSTOM. Set it to 0 if invalid."</field>
+      <field type="uint8_t" name="signal_quality" units="%">Signal quality of the sensor. Specific to each sensor type, representing the relation of the signal strength with the target reflectivity, distance, size or aspect, but normalised as a percentage. 0 = unknown/unset signal quality, 1 = invalid signal, 100 = perfect signal.</field>
```
Comment (tridge) [inline message_definitions/v1.0/common.xml:4717]: why can't we use the covariance?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by khancyr** [comment]
do we have an use case for this ?

File: message_definitions/v1.0/common.xml (hunk 1 above)
Comment (amilcarlucas) [reply message_definitions/v1.0/common.xml:4717]: @hamishwillee can you comment on tridge?

Comment (hamishwillee) [reply message_definitions/v1.0/common.xml:4717]: Nope. Perhaps @auturgy can remember this discussion?

Comment (hamishwillee) [reply message_definitions/v1.0/common.xml:4717]: @julian.oes Can you remember this? I can dig, but probably you have a better view.

**Comment by amilcarlucas** [comment]
@julianoes can you comment on this?

**Comment by julianoes** [comment]
I vaguely remember being confused by this. It was added by @TSC21 in https://github.com/mavlink/mavlink/pull/1403.

**Comment by amilcarlucas** [comment]
So according to @TSC21 this is something that some Lidars like the RPlidar A2 provide. Do we want this on ArduPilot?

**Comment by tridge** [comment]
> So according to @TSC21 this is something that some Lidars like the RPlidar A2 provide. Do we want this on ArduPilot?

so why doesn't that driver map it to a covariance?
we'd be paying a very high price for this feature in bandwidth

**Comment by TSC21** [comment]
> > So according to @TSC21 this is something that some Lidars like the RPlidar A2 provide. Do we want this on ArduPilot?
> 
> so why doesn't that driver map it to a covariance?
> we'd be paying a very high price for this feature in bandwidth

Because mapping between the signal quality (related to signal metrics that depend on several different factors) and covariance (should be variance, related to data uncertainty) is not obvious and the relation might not even be provided by the sensor datasheet. For example, on the Lidar Lite v3 for example, you can get the signal strength value, and you can read on their datasheet how it is directly or indirectly affected by distance, target size, aspect and reflectivity. Knowing the peak value of the signal, you can normalize the signal strength value and use it as a metric for sensor data validation, paired or not with the covariance.

Can you explain why would a unsigned int extension, defaulted to 0, would affect the bandwidth that much?

**Comment by hamishwillee** [comment]
An default zero value extension should not affect the bandwidth at all on systems that choose not to use it.

File: message_definitions/v1.0/common.xml (hunk 1 above)
Comment (amilcarlucas) [reply message_definitions/v1.0/common.xml:4187]: I guess the issue here is not the overhead when signal_quality = 0.
The problem is the overhead when signal_quality != 0 and (horizontal_fov = 0 and vertical_fov = 0 and quaternion = {0} )
Is this your point here @tridge ?
@TSC21 is it worth the overhead?

Comment (TSC21) [reply message_definitions/v1.0/common.xml:4187]: I would say that sensor that provides you a signal quality metric that you can use, will usually also provide you hfov and vfov. If it's worth? Depend on how much detailed your sensor data to be.

**Comment by tridge** [comment]
we can just decide not to fill in this field

**Review by tridge — APPROVED** [review]

**Pull request merged**